src/blockchain/alt_block_tree.cpp

```diff
@@ -46,259 +46,91 @@ bool AltTree::bootstrap(ValidationState& state) {
+    // ensure ids are unique within `pop`
+    bool inserted = ids.insert(id).second;
+    return !inserted;
+  });
 
-  return true;
+  return newend;
 }
 
-// check for Payload duplicates in ancestor and descendant blocks
-template <typename P>
-bool checkNoPayloadDuplicatesInOtherBlocks(AltTree& tree,
-                                           AltTree::index_t& index,
-                                           std::vector<P>& payloads,
-                                           ValidationState& state,
-                                           bool continueOnInvalid) {
-  auto& storage = tree.getStorage();
-
-  for (auto it = payloads.begin(); it != payloads.end();) {
-    auto& payload = *it;
-    auto pid = payload.getId();
-    bool failedPayload = false;
-
-    for (const auto& containingBlock :
-         storage.getContainingAltBlocks(pid.asVector())) {
-      auto* containingIndex = tree.getBlockIndex(containingBlock);
-      VBK_ASSERT(
-          containingIndex != nullptr &&
-          "state corruption: the storage index and block tree are out of sync");
-
-      // is `containing` block of this payload is descendant of `index`?
-      if (containingIndex->getHeight() > index.getHeight()) {
-        if (containingIndex->getAncestor(index.getHeight()) == &index) {
-          // we called AddPayloads(index, pid), where `index` is on the same
-          // chain as `containing` and `containing` already has payload with id
-          // `pid`. Block that is later, becomes invalid, all its descendants
-          // become invalid.
-          storage.setValidity(containingBlock, pid, false);
-          tree.invalidateSubtree(
-              *containingIndex, BLOCK_FAILED_POP, /*do fr=*/false);
-        } else {
-          // `containing` already has `pid`, but is on different chain than
-          // `index`
-        }
-      } else {
-        // check if this is an ancestor that contains the `pid`
-        if (index.getAncestor(containingIndex->getHeight()) ==
-            containingIndex) {
-          if (continueOnInvalid) {
-            failedPayload = true;
-            break;
-          }
-
-          return state.Invalid(
-              "ALT-duplicate-payloads-" + P::name() + "-ancestor",
-              fmt::sprintf("Ancestor block=%s already contains %s=%s that we "
-                           "attempted to add to block=%s.",
-                           containingIndex->toPrettyString(),
-                           P::name(),
-                           payload.toPrettyString(),
-                           index.toPrettyString()));
-        }
-      }  // end if
-    }    // end for
-
-    it = failedPayload ? payloads.erase(it) : ++it;
-  }  // end for
+template <typename Pop>
+bool searchForDuplicates(BlockIndex<AltBlock>& index,
+                         std::vector<Pop>& pop,
+                         AltTree& tree,
+                         ValidationState& state,
+                         bool continueOnInvalid) {
```

Comment:
It is not just duplicates search but also Pop filtering. Are you sure you want to modify pop structure?

Reply:
Yes. It is used in mempool.